ultravox/tools/extract_hubert.py

```diff
@@ -0,0 +1,119 @@
+    if save_path.exists():
+        logger.debug(f"Skipping download of existing file [file={save_path}]")
+        return
+
+    logger.debug(f"Downloading [file={save_path}]")
+    response = requests.get(
+        f"https://dl.fbaipublicfiles.com/hubert/{filename}", stream=True
+    )
+    response.raise_for_status()
+    with open(save_path, "wb") as f:
+        for chunk in response.iter_content(chunk_size=8192):
+            f.write(chunk)
+
+
+def _ensure_models_present():
+    _MODEL_DIR.mkdir(parents=True, exist_ok=True)
+    _download_file(_MODEL_CHECKPOINT)
+    _download_file(_MODEL_KMEANS)
+    _download_file(_MODEL_CHECKPOINT_MULTI)
+    _download_file(_MODEL_KMEANS_MULTI)
+
+
+def _load_model(multilingual):
+    _ensure_models_present()
+    if multilingual:
+        checkpoint_path = _MODEL_DIR / _MODEL_CHECKPOINT_MULTI
+        kmeans_path = _MODEL_DIR / _MODEL_KMEANS_MULTI
+        return HubertWithKmeans(
+            checkpoint_path.as_posix(),
+            kmeans_path.as_posix(),
+            output_layer=_MULTILINGUAL_OUTPUT_LAYER,
+        )
+    else:
+        checkpoint_path = _MODEL_DIR / _MODEL_CHECKPOINT
+        kmeans_path = _MODEL_DIR / _MODEL_KMEANS
+        return HubertWithKmeans(checkpoint_path.as_posix(), kmeans_path.as_posix())
+
+
+def _replace_suffix(filename, new_suffix):
+    return filename.with_name(filename.stem + new_suffix)
+
+
+def _process_file(device, hubert, filename):
+    try:
+        new_suffix = "_hubert.npy" if not args.multilingual else "_mhubert.npy"
+        output_filename = _replace_suffix(filename, new_suffix)
+        if output_filename.exists():
+            logger.debug(f"Skipping pre-extracted file [file={output_filename}].")
+            return
+
+        audio, sr = load(filename)
+        audio = resample(audio.to(device), sr, _SAMPLE_RATE)
+
+        tokens = hubert(audio)
+        tokens = tokens.squeeze(0).cpu().numpy()
+
+        logger.debug(f"Saving [file={output_filename}, token_len={len(tokens)}].")
+        np.save(output_filename, tokens)
+    except Exception as e:
+        logger.warning(f"Failed to process [file={filename}, error={e}]")
+
+
+def _list_supported_audio_files(directory, extensions=_AUDIO_SUFFIXES):
+    return [file for ext in extensions for file in Path(directory).rglob(ext)]
+
+
+def main(args):
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    logger.info("Loading model.")
+    hubert = _load_model(args.multilingual).to(device)
+    logger.info("Model loaded.")
+    # TODO(shaper): Could parallelize with `GPUExecutor` or similar.
+    audio_files = _list_supported_audio_files(args.data_dir)
+    logger.info(f"Found {len(audio_files)} audio files.")
+    for filename in tqdm(
+        audio_files, desc="Extracting Hubert tokens", total=len(audio_files)
+    ):
+        _process_file(device, hubert, filename)
+
+
+if __name__ == "__main__":
+    parser = ArgumentParser()
+    parser.add_argument("data_dir")
```

Comment:
I'm not fully opposed to offline creation of the dataset, but we'll definitely have to balance the pros and cons. I see 3 options generally for this:

1. online streaming (which we currently use in Ultravox)
2. offline done once to update dataset
3. offline every time at the start of training

I'm not sure if you had 2 or 3 in mind.
- Option 3 doesn't make much sense to me (at least as slow as 1 and causes training runs to be started with a huge delay).
- Option 2 is mainly good when you are fairly certain which encoder you want to use, which IMO is still up to debate at this point.
- Option 1 gives you the most flexibility. If you have any data that shows the costs for this is too high I can be convinced to give up that flexibility

Reply:
This work is still early and experimental for sure. I had in mind to land the experimental tools and such as we go, presuming they get to some point where it seems likely they'll be of at least semi lasting value. However, if we are thinking of this repo as more production-ready oriented work or want to separate things that are still super early into other subdirectories, etc., I am happy to shift my thinking.

Here I had 2 in mind, though in our experimentation we've also done 1. It did appear to slow training, I want to say fairly substantially, but I'll look at it again and follow up.

In general where we are today is we're looking to effectively update various datasets to add HuBERT tokens (and I imagine SNAC and others as well over time) and it seems advantageous from an overall compute perspective to have updated datasets with those tokens. Even if they live only locally with folks experimenting, vs. pushed to HF or otherwise, having a tool to allow doing this seems useful. We are using it today e.g. I used it for AnyInstruct and am looking now at what other datasets could be most useful for experimenting with, and will apply it to those as well.

Reply:
Okay makes sense. Assuming that you're treating this as an experimental feature and the original path still remains viable then I'm more than fine with this.